### DISCUSSION TITLE
py版本 ： Python 3.12.6  。在执行裁剪并添加字幕操作后报错

### DISCUSSION THREAD

**Discussion by Steamedyu** [modelscope/FunClip#115] [Q&A] [answered]
没有看见readme中的\config_defaults.py文件，只有config并修正路径IMAGEMAGICK_BINARY = os.getenv("C:\Program Files\ImageMagick-7.1.1-Q16\magick.exe", "auto-detect")
--------------------------------------------------------------------------------
报错信息
AttributeError: 'SubtitlesClip' object has no attribute 'set_pos'. Did you mean: 'set_fps'?
Traceback (most recent call last):
  File "D:\开发\py\Lib\site-packages\gradio\queueing.py", line 536, in process_events
    response = await route_utils.call_process_api(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\开发\py\Lib\site-packages\gradio\route_utils.py", line 321, in call_process_api
    output = await app.get_blocks().process_api(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\开发\py\Lib\site-packages\gradio\blocks.py", line 1935, in process_api
    result = await self.call_function(
             ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\开发\py\Lib\site-packages\gradio\blocks.py", line 1520, in call_function
    prediction = await anyio.to_thread.run_sync(  # type: ignore
    result = await self.call_function(
             ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\开发\py\Lib\site-packages\gradio\blocks.py", line 1520, in call_function
    prediction = await anyio.to_thread.run_sync(  # type: ignore
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\开发\py\Lib\site-packages\anyio\to_thread.py", line 56, in run_sync
             ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\开发\py\Lib\site-packages\gradio\blocks.py", line 1520, in call_function
    prediction = await anyio.to_thread.run_sync(  # type: ignore
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\开发\py\Lib\site-packages\anyio\to_thread.py", line 56, in run_sync
  File "D:\开发\py\Lib\site-packages\gradio\blocks.py", line 1520, in call_function

**Comment by LauraGPT** [comment, marked as answer]
这个报错不是 Python 3.12 本身导致的，而是安装到了 **MoviePy 2.x**，与 FunClip 当前使用的 MoviePy 1.x API 不兼容。

两个现象都能确认这一点：

- MoviePy 2.x 不再提供 `moviepy/config_defaults.py`
- MoviePy 2.x 把 `set_*` API 改成了 `with_*`，所以 `SubtitlesClip.set_pos` 不存在

FunClip 当前主干的 `requirements.txt` 已固定 `moviepy==1.0.3`。建议在新的虚拟环境里完整安装仓库依赖：

```powershell
py -3.12 -m venv .venv
.\.venv\Scripts\Activate.ps1
python -m pip install --upgrade pip
python -m pip install -r requirements.txt
python -c "import moviepy; print(moviepy.__version__)"
```

最后一条应输出：

```text
1.0.3
```

然后在**启动 FunClip 前**设置 ImageMagick 路径：

```powershell
$env:IMAGEMAGICK_BINARY = "C:\Program Files\ImageMagick-7.1.1-Q16\magick.exe"
python funclip/launch.py
```

如果仍选择修改 MoviePy 1.0.3 的 `config_defaults.py`，应直接写原始字符串：

```python
IMAGEMAGICK_BINARY = r"C:\Program Files\ImageMagick-7.1.1-Q16\magick.exe"
```

不能写成 `os.getenv("C:\Program Files\...", "auto-detect")`；`os.getenv` 的第一个参数是**环境变量名**，不是可执行文件路径。

也不建议只把 FunClip 中的 `set_pos` 改成 `with_position`：MoviePy 2.x 还删除了 `moviepy.editor` 并更改了 `TextClip` 等接口，需要一次完整迁移。当前稳定方案是使用仓库锁定的 1.0.3。参考：[FunClip requirements](https://github.com/modelscope/FunClip/blob/80ea6ea18b8937db0d453e3f5440fdfbc4778a32/requirements.txt#L7) 和 [MoviePy 官方 v2 迁移说明](https://zulko.github.io/moviepy/getting_started/updating_to_v2.html)。
